tests/test_validate_data.py

```diff
@@ -224,11 +225,41 @@ def test_DataValidator_validate_fail_with_warning(file, value, simple_df, caplog
 
     with pytest.raises(ValueError, match="Data validation failed"):
         data_validator.apply(simple_df)
-        
+
     # check if the log message contains the correct information
     assert failed_validation_message in caplog.text
 
 
+def test_DataValidator_pass_with_warning(simple_df, caplog):
+
+    data_validator = DataValidator.from_file(
+        DATA_VALIDATION_TEST_DIR / "validate_pass_warning.yaml"
+    )
+    warning_message = (
+        "Data validation with warning(s) "
+        f"""(file {(DATA_VALIDATION_TEST_DIR / "validate_pass_warning.yaml").relative_to(Path.cwd())}):
+  Criteria: variable: ['Primary Energy'], upper_bound: 6.5
+       model scenario region        variable   unit  year  value warning_level
+  0  model_a   scen_b  World  Primary Energy  EJ/yr  2010    7.0          high
+
+  Criteria: variable: ['Primary Energy'], upper_bound: 1.5
+       model scenario region        variable   unit  year  value warning_level
+  0  model_a   scen_a  World  Primary Energy  EJ/yr  2010    6.0           low
+  1  model_a   scen_b  World  Primary Energy  EJ/yr  2005    2.0           low"""
+    )
+
+    data_validator.apply(simple_df)
```

Comment:
Even though it technically works because we're modifying `simple_df` in place, I'd suggest the following:
```suggestion
    obs = data_validator.apply(simple_df)
```
We could think about whether or not we want to allow `Processor`s to modify the input to `.apply()` in place anyway rather than return a modified deep copy but I guess that's a conversation for another day. For what it's worth, my preference lies strongly with not touching the input data and rather returning a deep copy. Reason being that for processors that are allowed to fail, we want to be able to retrieve the unmodified input data to continue.

Reply:
I see why letting `apply()` return a copy instead of the actual (modified) **IamDataFrame** is appealing, but I'm worried about performance with large scenario submissions and we run a dozen **Processor** instances in sequence and create a copy of the data for each...

Reply:
Of course, no such thing as a free lunch. Let's just hope that we don't get bitten by this design at some point. Or at least that we quickly realize what's going on.